File: leetcode/leetcode876.cpp
#include <iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode()
        : val(0)
        , next(nullptr)
    {
    }
    ListNode(int x)
        : val(x)
        , next(nullptr)
    {
    }
    ListNode(int x, ListNode* next)
        : val(x)
        , next(next)
    {
    }
};

class Solution {
public:
    ListNode* middleNode(ListNode* head)
    {
        int sz = 0;
        ListNode* n = head;
        while (n != nullptr) {
            sz++;
            n = n->next;
        }
        n = head;
        sz /= 2;
        while (sz--) {
            n = n->next;
        }
        return n;
    }
};
int main(void)
{
    return 0;
}

File: leetcode/100-999/leetcode876.py
# Definition for singly-linked list.
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        slow = fast = head
        while fast.next:
            slow = slow.next
            fast = fast.next
            if not fast.next:
                break
            fast = fast.next
        return slow